OpenSim/Actuators/VandenBogert2011Muscle.cpp

```diff
@@ -0,0 +1,1234 @@
+const
+//TODO: Add clamping?  Add ignore_activation_dynamics
+{
+    setStateVariableValue(s, "activation", activation);
+
+    //markCacheVariableInvalid(s,"velInfo");
+    //markCacheVariableInvalid(s,"dynamicsInfo");
+}
+
+// Get the Residual of muscle
+// TODO:  This breaks naming convention in that it does not get an already
+// evaluated constant, it actually performs the calculation.  Without using a
+// state variable or adding to the cache variables, I think I have to do it this
+// way?
+SimTK::Vec2 VandenBogert2011Muscle::getResidual(const SimTK::State& s,
+                                                double projFibVelNorm_guess,
+                                                double activdot_guess, double u)
+                                                const
+{
+    ImplicitResidual residualStruct = calcImplicitResidual(s,
+                                                           projFibVelNorm_guess,
+                                                           activdot_guess,
+                                                           u, 0);
+
+    SimTK::Vec2 residual = {residualStruct.forceResidual,
+                            residualStruct.activResidual};
+    return(residual);
+}
+
+
+
+//These are hacks because I am not using muscle.h mli cache variables, yet:
+double  VandenBogert2011Muscle::getFiberLength(SimTK::State& s) const
+{
+    double fiberLength=getStateVariableValue(s, "fiber_length");
+    return fiberLength;
+}
+double  VandenBogert2011Muscle::getActivation(SimTK::State& s) const
+{
+    double activ=getStateVariableValue(s, "activation");
+    return activ;
+}
+
+
+
+//==============================================================================
+// Muscle.h Interface
+//==============================================================================
+
+
+double  VandenBogert2011Muscle::computeActuation(const SimTK::State& s) const
+{
+    double activ=getActivation(s);
+    // For now we will start with static guess
+    // TODO: figure out how to use previous state as guess
+    SimTK::Vector ydotInitialGuess(2);
+    ydotInitialGuess[0]=0.0;
+    ydotInitialGuess[1]=0.0;
+    SimTK::Vec2 sdot=calcSolveMuscle(s, activ, ydotInitialGuess);
+
+    VandenBogert2011Muscle::ImplicitResidual Result= calcImplicitResidual(s,
+                                          sdot[0],
+                                          sdot[1],
+                                          activ,
+                                          0);
+
+    //TODO: Could probably add a check here that residuals =0
+    double actuation=Result.forceTendon;
+
+    return(actuation);
+}
+
+
+
+void VandenBogert2011Muscle::computeInitialFiberEquilibrium(SimTK::State& s)
+const
+{
+    // TODO:  this is the same code as computeFiberEquilibriumAtZeroVelocity
+    //Need to update for v~=0
+    //Let's start with the default activation and the fiberVelocity=0
+
+    computeFiberEquilibriumAtZeroVelocity(s);
+
```

Comment:
I just figured out what this is supposed to do, so I'm going to write it even if you already figured it out. The method that Millard uses for this is described in the appendix of the 2013 paper. You're solving for both fiber length and fiber velocity.  The two equations are the force equilibrium `fiberForce - tendonForce = 0` and an approximation of its derivative to help you determine what the fiber velocity should be; equation (A6) in the paper:

![image](https://cloud.githubusercontent.com/assets/846001/17556226/87d2db34-5ec7-11e6-8b2b-f0f7903eb2a6.png)

I think it would also be possible to use the exact derivative of the force equilibrium equation (with respect to time) as the additional constraint, since you mostly have that via the exact jacobian (`dResidual_dt = dResidual_dFibLength * dFibLength_dt + dResidual_dFibVel * dFibVel_dt + dResidual_dActiv * dActiv_dt + dResidual_dMTLength * dMTLength_dt`), but I think this is needlessly complex for a constraint that only needs to be approximate.

@aseth1 or @tkuchida probably can lend better insight.